### PR TITLE
Use attrs-based objects to hold results of explain_weights and explain_prediction

### PR DESCRIPTION
**Pull request by lopuhin**
This is a purely mechanical change apart from one thing: we had distinct "classes" and "targets" keys for classification and regression, and "class"/"target" keys inside them - in this branch I changed to use "targets" in all cases, because it seems applicable to classification too, and simplified code and API (this is similar to what we did with "estimator" key).

Right now I think that changing from dicts to objects is a good thing to do, because of:
- the API will be more future-proof - we can deprecate stuff, add properties, etc.
- potentially, IDE support is better (~~right now PyCharm does not understand attrs classes, but we can fix it by adding ``__init__`` to them~~, and also add type annotations)
- we can also have more benefit from mypy-lang and typing in the future
- we can catch errors where we mistype an attribute name more easily (here we declare up-front what attributes does an object have)

We can still convert objects to dicts using ``attr.asdict`` function (it is recursive by default).

The only downside I see so far is that ``__repr__`` of attrs objects contains the class name and all attribute names and values, but they are not formatted as nicely as IPython can do it (there are no line breaks added). But we can disable this custom ``__repr__`` or add our own.

There are no new features here - the branch touches a lot of code, so to minimize conflicts it will be easier to merge it first and then add new features.

### CODE REVIEW

**Comment by lopuhin** [comment]
Forgot to say - I additionally checked html output with master, there were only a couple of differences in the number of negative and positive features in newsgroup tests, probably there is still some non-deterministic behaviour left in scikit.

File: eli5/base.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+
+import attr
+from typing import Dict, List, Tuple, Union
```
Comment (kmike) [inline eli5/base.py]: what do you think about swapping imports? I always think about typing as an stdlib module

File: eli5/base.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+@attr.s
+class Explanation(object):
```
Comment (kmike) [inline eli5/base.py]: Could you please add docstrings to all these classes? They don't have to be long; maybe you can move some of comments from explain_.. functions here as well.

Comment (lopuhin) [inline eli5/base.py]: Right, docstrings and some comment would be definitely nice here, let me add them tomorrow :)

Comment (kmike) [inline eli5/base.py]: Let's do it in another PR; this one looks good as-is!
